File: 0278-first-bad-version/0278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 0;
        int end = n;
        int solution = n;
        while(start < end){
            int mid = start + (end-start)/2;
            boolean isBad = isBadVersion(mid);
            if(isBad){
                end = mid;
                solution = Math.min(mid, solution);
            } else {
                start = mid + 1;
            }
        }
        return solution;
    }
}

File: 0278-first-bad-version/0278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        start = 0
        last = n
        solution = n
        while start < last:
            mid = (start + last)//2
            isBad = isBadVersion(mid)
            if isBad:
                last = mid
                solution = min(solution, mid)
            else:
                start = mid + 1
        return solution
        